src/rdkafka_request.c

```diff
@@ -2619,7 +2619,12 @@ static void rd_kafka_handle_Metadata(rd_kafka_t *rk,
                                       RD_KAFKA_ERR_ACTION_END);
 
         if (actions & RD_KAFKA_ERR_ACTION_RETRY) {
-                if (rd_kafka_buf_retry(rkb, request))
+                /* In case it's a brokers full refresh call,
+                 * avoid retrying it on this same broker.
+                 * This is to prevent client is hung
+                 * until it can connect to this broker again. */
+                if (!request->rkbuf_u.Metadata.decr &&
```

Comment:
Wouldn't we need to hold 
```
Metadata.decr_lock
```
while reading it?

Reply:
No, it's only needed if we need to decrease the integer it points to, that can be `rkmc_full_brokers_sent` or `rkmc_full_topics_sent`. Lock for it will be acquired when the buffer is destroyed.